states/snake.mod.py

```diff
@@ -1,168 +1,192 @@
-                else:
-                    for pixel in self.game.apples:
-                        pixels[pixel[1]][pixel[0]] = RED
-                    for pixel in self.game.snake:
-                        pixels[pixel[1]][pixel[0]] = self.code
-
-                self.output_frame(bytes(self.flatten(pixels)))
-                if self.game.dead:
-                    sleep(2)
-                    self.game.reset()
-
-                if not self.on_pi:
-                    time_end = time()
-                    time_delta = time_start - time_end
-                    sleep(max(0.04-time_delta, 0))
+
+class Game:
+    def __init__(self):
+        self.players = []
+        self.apples = []
+        self.bodies = []
+        threading.Thread(target=self.receive).start()
+
+    @staticmethod
+    def hex_to_rgb(code):
+        return tuple(int(code[i:i + 2], 16) for i in (0, 2, 4))
+
+    def update(self, tick):
+        for player in self.players:
+            if not player.active:
+                continue
+            speed = 30 - round(len(player.elements) / 2)
+            if speed < 5:
+                speed = 5
+
+            if tick % speed == 0:
+                player.update()
+                for p2 in self.players:
+                    for index, element in enumerate(p2.elements):
+                        if p2 == player and index == 0:
+                            continue
+                        if element == player.elements[0]:
+                            player.dead = True
+
+                if player.dead:
+                    self.bodies.append(Body(self.hex_to_rgb(player.color), player.elements))
+                    player.reset()
+
+            for apple in self.apples:
+                if player.elements[0] == apple.location:
+                    player.size += 1
+                    apple.newLoc()
+
+        for body in self.bodies:
+            body.fade()
+
+    def export(self):
+        image = Image.new("RGB", DIMENSIONS, "black")
+        draw = ImageDraw.Draw(image)
+        for body in self.bodies:
+            for element in body.elements:
+                draw.point(element, fill=body.color)
+        for player in self.players:
+            if not player.active:
+                continue
+            for element in player.elements:
+                draw.point(element, fill=self.hex_to_rgb(player.color))
+        for apple in self.apples:
+            draw.point(apple.location, fill=(255, 0, 0))
+        draw.rectangle([(0, 0), (DIMENSIONS[0] - 1, DIMENSIONS[1] - 1)], outline=(250, 128, 114))
+
+        return image
 
     def receive(self):
         with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
             s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
             s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
             s.bind(("0.0.0.0", 1029))
             s.listen()
-            while not self.killed:
+            while True:
```

Comment:
Hoe stopt deze thread? Socket exception?

Reply:
Niet, hier luistert hij naar nieuwe verbindingen, dat moet aan blijven

Reply:
Oei, dat gaat dan fout. De hele state class wordt namelijk gestopt/freed en opnieuw gestart na elke 'run' (dus elke keer nadat de state actief is geweest). Als deze thread blijft draaien blijft die poort in gebruik en krijg je een exception van de nieuwe thread dat de poort al bezet is. Dit zou dus een while not self.killed moeten zijn (wat er al stond)

Reply:
Oh, ik los dat dan zo op...

Reply:
Fixed